File: frontend/src/types/userCharacter/userCharacter.ts
// types/userCharacter/userCharacter.ts
export interface UserCharacter {
  id: number;
  user_id: number;

  // Character identifiers
  character_id: number;
  name: string;
  realm: string;
  region: string;

  // Basic information
  class: string;
  race: string;
  gender: string;
  faction: string;
  active_spec_name: string;
  active_spec_id: number;
  active_spec_role: string;

  // Level and main stats
  level: number;
  item_level: number;
  mythic_plus_rating: number;
  mythic_plus_rating_color: string;
  achievement_points: number;
  honorable_kills: number;

  // Image URLs
  avatar_url: string;
  inset_avatar_url: string;
  main_raw_url: string;
  profile_url: string;

  // JSON structured data (optional for now)
  equipment_json?: Record<string, any>;
  stats_json?: Record<string, any>;
  talents_json?: Record<string, any>;
  mythic_plus_json?: Record<string, any>;
  raids_json?: Record<string, any>;

  // Metadata
  is_displayed: boolean;
  last_api_update: string;
  created_at: string;
  updated_at: string;
}

// CharacterBasicInfo contains basic informations about a character
export interface CharacterBasicInfo {
  character_id: number;
  name: string;
  realm: string;
  region: string;
  class: string;
  race: string;
  level: number;
  faction: string;
}

// SyncResult represents the result of a sync
export interface SyncResult {
  message: string;
  count: number;
}

// RefreshResult represents the result of a refresh
export interface RefreshResult {
  message: string;
  new_characters: number;
  updated_characters: number;
}

// WoWProfile is the profile of the user
export interface WoWProfile {
  sub: string;
  id: number;
  region?: string;
  _links: {
    self: {
      href: string;
    };
  };
  wow_accounts: Array<{
    id: number;
    characters: Array<{
      id: number;
      name: string;
      realm: {
        id: number;
        slug: string;
        name: string;
      };
      playable_class: {
        id: number;
        name: string;
      };
      playable_race: {
        id: number;
        name: string;
      };
      level: number;
      faction: {
        type: string;
        name: string;
      };
    }>;
  }>;
}

// CharacterDisplayResponse is the response for the character display endpoints
export interface CharacterDisplayResponse {
  message: string;
}

export interface FavoriteCharacterResponse {
  message: string;
}

// Enum centralised for WoW error codes
export enum WoWErrorCode {
  UNAUTHORIZED = "wow_unauthorized",
  TOKEN_EXPIRED = "wow_token_expired",
  NETWORK_ERROR = "wow_network_error",
  BATTLE_NET_NOT_LINKED = "battle_net_not_linked",
  NOT_FOUND = "wow_not_found",
  SERVER_ERROR = "wow_server_error",
}

// Centralised error class
export class WoWError extends Error {
  constructor(
    public code: WoWErrorCode,
    message: string,
    public originalError?: unknown
  ) {
    super(message);
    this.name = "WoWError";
  }
}

File: backend/internal/models/user_character.go
package models

import (
	"time"

	"gorm.io/datatypes"
	"gorm.io/gorm"
	"gorm.io/gorm/clause"
)

// UserCharacter represents a WoW character linked to a user account
type UserCharacter struct {
	gorm.Model
	ID     uint `gorm:"primaryKey" json:"id"`
	UserID uint `gorm:"index;not null" json:"user_id"`

	// Character identifiers
	CharacterID int64  `gorm:"index;not null" json:"character_id"`
	Name        string `gorm:"not null" json:"name"`
	Realm       string `gorm:"not null" json:"realm"`
	Region      string `gorm:"not null" json:"region"`

	// Basic information
	Class          string `gorm:"not null" json:"class"`
	Race           string `gorm:"not null" json:"race"`
	Gender         string `gorm:"not null" json:"gender"`
	Faction        string `gorm:"not null" json:"faction"`
	ActiveSpecName string `gorm:"not null" json:"active_spec_name"`
	ActiveSpecID   int    `gorm:"not null" json:"active_spec_id"`
	ActiveSpecRole string `gorm:"not null" json:"active_spec_role"`

	// Level and main stats
	Level                 int     `json:"level"`
	ItemLevel             float64 `json:"item_level"`
	MythicPlusRating      float64 `json:"mythic_plus_rating"`
	MythicPlusRatingColor string  `json:"mythic_plus_rating_color"`
	AchievementPoints     int     `json:"achievement_points"`
	HonorableKills        int     `json:"honorable_kills"`

	// Image URLs
	AvatarURL      string `json:"avatar_url"`
	InsetAvatarURL string `json:"inset_avatar_url"`
	MainRawURL     string `json:"main_raw_url"`
	ProfileURL     string `json:"profile_url"`

	// JSON structured data
	EquipmentJSON  datatypes.JSON `gorm:"type:jsonb" json:"equipment_json,omitempty"`
	StatsJSON      datatypes.JSON `gorm:"type:jsonb" json:"stats_json,omitempty"`
	TalentsJSON    datatypes.JSON `gorm:"type:jsonb" json:"talents_json,omitempty"`
	MythicPlusJSON datatypes.JSON `gorm:"type:jsonb" json:"mythic_plus_json,omitempty"`
	RaidsJSON      datatypes.JSON `gorm:"type:jsonb" json:"raids_json,omitempty"`

	// Metadata
	IsDisplayed   bool      `gorm:"default:true" json:"is_displayed"`
	LastAPIUpdate time.Time `json:"last_api_update"`
}

// TableName overrides the table name
func (UserCharacter) TableName() string {
	return "user_characters"
}

// BeforeCreate ensures uniqueness of character across realm and region
func (uc *UserCharacter) BeforeCreate(tx *gorm.DB) error {
	// Set unique composite index on character_id, realm, region
	tx.Statement.AddClause(clause.OnConflict{
		Columns:   []clause.Column{{Name: "character_id"}, {Name: "realm"}, {Name: "region"}},
		DoNothing: true,
	})
	return nil
}
